### ISSUE TITLE
[BUG REPORT] Traversing through dataset data records is buggy

### ISSUE THREAD

**Issue by hhsecond** [tensorwerk/hangar-py#49]
**Describe the bug**
The function `RecordQuery._traverse_dataset_data_records(self, dataset_name)` is has a critical bug which related to the condition check we are doing to make sure we are checking for the provided dataset (This must be same with `_traverse_dataset_schema_records` function as well since the logic is similar there as well). If the repository has multiple datasets where one's name has another dataset's name as a substring in the start of the name, would cause the condition check `dataRecKey.startswith(startDatasetRecordCountRangeKey)` to return wrong data.

**Severity**

Select an option:
- [ ] Data Corruption / Loss of Any Kind
- [x] Unexpected Behavior, Exceptions or Error Thrown
- [ ] Performance Bottleneck

**To Reproduce**
First code snippet to save two datasets in the same repository named `MNIST` and `MNISTLabel`

```python
import os
from hangar import Repository
import torchvision
import numpy as np

data = torchvision.datasets.MNIST('.', download=True)
path = os.path.dirname(os.path.abspath(__file__))
repo = Repository(path=path)
repo.init(user_name='hhsecond', user_email='sherin@tensorwerk.com', remove_old=True)
co = repo.checkout(write=True)
co.datasets.init_dataset('MNIST', shape=(28, 28), dtype=np.uint8)
co.datasets.init_dataset('MNISTLabel', shape=(1,), dtype=np.int64)
mnist_dset = co.datasets['MNIST']
label_dset = co.datasets['MNISTLabel']

with mnist_dset, label_dset:
    for i, (image, label) in enumerate(data):
        print(i)
        mnist_dset[str(i)] = np.array(image)
        label_dset[str(i)] = np.array([label])
co.commit('0.1')
co.close()
```

The second part where we could try to check the name of dataset `MNIST`
```python
co = repo.checkout()
mnist = co.datasets['MNIST']
mnistlabel = co.datasets['MNISTLabel']
print(len(mnist))
```
`len` on dataset triggers the above functions which would cause it to fail

**Expected behavior**
- In the above case, the user should get the actual length
- Generally, `_traverse_dataset_data_records` should return a dictionary of data records with name as key and hash as value
